### PR TITLE
L1T Phase-2 Tk objects - to 12_4_X

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

Same as https://github.com/cms-sw/cmssw/pull/37189 but to 12_4_X and with @srimanob comments included. 

This is the first PR to port all L1T Phase-2 developments to cms-sw. This first PR targets the Phase-2 Tk objects. It includes several PRs made locally, the details of which can be found in the twiki https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideL1TPhase2Instructions and in the cms-l1t-offline branch.

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
@srimanob Your comments to the PR in 12_3 have been addressed in this new PR. Can you please sign off if everything looks good?

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
+Upgrade

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 1]
```diff
@@ -161,9 +163,9 @@ void L1FastTrackingJetProducer::produce(edm::Event& iEvent, const edm::EventSetu
         theStubs = iterL1Track->getStubRefs();
     int trk_nstub = (int)theStubs.size();
 
-    if (std::abs(trk_z0) > trkZMax_)
+    if (fabs(trk_z0) > trkZMax_)
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:166]: Why reverting back from `std::abs` to `fabs`? (Here and below)
We normally require to do the opposite...

File: L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        << "minpt: " << jetMinPt_ << " maxeta: " << jetMaxEta_ << " minNtracksHighPt: " << minNtracksHighPt_
+        << " minNtracksLowPt: " << minNtracksLowPt_ << "\n"
+        << "====CUT AP_INTS TO FLOATS====\n"
+        << "minpt: " << (float)jetMinPt_ * l1tmhtemu::kStepPt << " maxeta: " << (float)jetMaxEta_ * l1tmhtemu::kStepEta
+        << " minNtracksHighPt: " << (int)minNtracksHighPt_ << " minNtracksLowPt: " << (int)minNtracksLowPt_ << "\n"
+        << "-------------------------------------------------------------------------\n";
+  }
+}
+
+L1TkHTMissEmulatorProducer::~L1TkHTMissEmulatorProducer() {}
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc:103]:
```suggestion
```

File: L1Trigger/L1TTrackMatch/plugins/L1TkHTMissProducer.cc [hunk 3]
```diff
@@ -76,8 +74,9 @@ void L1TkHTMissProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSet
   std::unique_ptr<TkHTMissCollection> MHTCollection(new TkHTMissCollection);
 
   // L1 primary vertex
-  edm::Handle<TkPrimaryVertexCollection> L1VertexHandle;
+  edm::Handle<VertexCollection> L1VertexHandle;
   iEvent.getByToken(pvToken_, L1VertexHandle);
+  //std::vector<Vertex>::const_iterator vtxIter;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TkHTMissProducer.cc:79]: Is this needed?

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 4]
```diff
@@ -231,35 +219,23 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
         float jetPy = jetPt * sin(jetPhi);
         float jetPz = jetPt * sinh(jetEta);
         float jetP = jetPt * cosh(jetEta);
-        int totalTighttrk_ = mzb.clusters[j].numttrks;
         int totalDisptrk = mzb.clusters[j].numtdtrks;
-        int totalTightDisptrk = mzb.clusters[j].numttdtrks;
-
+        bool isDispJet = false;
+        if (totalDisptrk > nDisplacedTracks_ || totalDisptrk == nDisplacedTracks_)
+          isDispJet = true;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:225]:
```suggestion
        bool isDispJet = totalDisptrk >= nDisplacedTracks_;
```

File: L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc [hunk 5]
```diff
@@ -32,41 +35,54 @@ class L1TrackerEtMissProducer : public edm::stream::EDProducer<> {
...
-//constructor//
+// constructor//
 L1TrackerEtMissProducer::L1TrackerEtMissProducer(const edm::ParameterSet& iConfig)
-    : pvToken_(consumes<TkPrimaryVertexCollection>(iConfig.getParameter<edm::InputTag>("L1VertexInputTag"))),
-      trackToken_(consumes<std::vector<TTTrack<Ref_Phase2TrackerDigi_> > >(
-          iConfig.getParameter<edm::InputTag>("L1TrackInputTag"))),
+    : pvToken_(consumes<L1VertexCollectionType>(iConfig.getParameter<edm::InputTag>("L1VertexInputTag"))),
+      trackToken_(consumes<L1TTTrackCollectionType>(iConfig.getParameter<edm::InputTag>("L1TrackInputTag"))),
       tTopoToken_(esConsumes<TrackerTopology, TrackerTopologyRcd>(edm::ESInputTag("", ""))) {
   maxZ0_ = (float)iConfig.getParameter<double>("maxZ0");
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc:82]: All these parameters should be more conveniently declared const in the class header, and initialized in the class initialization list instead

File: L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc [hunk 6]
```diff
@@ -162,23 +194,20 @@ void L1TrackerEtMissProducer::produce(edm::Event& iEvent, const edm::EventSetup&
...
-      else if (fabs(eta) >= 1.0 && fabs(eta) < 1.2)
-        deltaZ_ = 0.76;
-      else if (fabs(eta) >= 1.2 && fabs(eta) < 1.6)
-        deltaZ_ = 1.0;
-      else if (fabs(eta) >= 1.6 && fabs(eta) < 2.0)
-        deltaZ_ = 1.7;
-      else if (fabs(eta) >= 2.0 && fabs(eta) <= 2.4)
-        deltaZ_ = 2.2;
+      for (unsigned int reg = 0; reg < etaRegions_.size(); reg++) {
+        if (fabs(eta) >= etaRegions_[reg] && fabs(eta) < etaRegions_[reg + 1]) {
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc:202]: This imply accessing `etaRegions_[etaRegions_.size()]`: the last bin should be treated differently

File: L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc [hunk 7]
```diff
@@ -190,20 +219,37 @@ void L1TrackerEtMissProducer::produce(edm::Event& iEvent, const edm::EventSetup&
   }  // end loop over tracks
 
   float et = sqrt(sumPx * sumPx + sumPy * sumPy);
-  double etmiss_PU = sqrt(sumPx_PU * sumPx_PU + sumPy_PU * sumPy_PU);
+  double etphi = atan2(sumPy, sumPx);
+  //double etmiss_PU = sqrt(sumPx_PU * sumPx_PU + sumPy_PU * sumPy_PU);
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissProducer.cc:223]: Remove the commented out code if not needed

File: L1Trigger/L1TTrackMatch/python/L1TrackSelectionProducer_cfi.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  useDisplacedTracksDeltaZOverride = cms.double(-1.0), # override the deltaZ cut value for displaced tracks
+  processSimulatedTracks = cms.bool(True), # return selected tracks after cutting on the floating point values
+  processEmulatedTracks = cms.bool(True), # return selected tracks after cutting on the bitwise emulated values
+  debug = cms.int32(0) # Verbosity levels: 0, 1, 2, 3, 4
+)
+
+L1TrackSelectionProducerExtended = L1TrackSelectionProducer.clone(
+  l1TracksInputTag = cms.InputTag("L1GTTInputProducerExtended","Level1TTTracksExtendedConverted"),
+  outputCollectionName = cms.string("Level1TTTracksExtendedSelected"),
+  useDisplacedTracksDeltaZOverride = cms.double(3.0), # Use promt/displaced tracks
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/python/L1TrackSelectionProducer_cfi.py:34]:
```suggestion
  l1TracksInputTag = "L1GTTInputProducerExtended : Level1TTTracksExtendedConverted",
  outputCollectionName = "Level1TTTracksExtendedSelected",
  useDisplacedTracksDeltaZOverride = 3.0, # Use prompt/displaced tracks
```

File: L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+#include "TPad.h"
+#include "TProfile.h"
+#include "TTree.h"
+
+using namespace std;
+
+class L1TkMETAnalyser : public edm::one::EDAnalyzer<> {
+public:
+  explicit L1TkMETAnalyser(const edm::ParameterSet& iConfig);
+  ~L1TkMETAnalyser() override;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc:40]:
```suggestion
  ~L1TkMETAnalyser() override = default;
```

File: L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  TrackMETEmuToken_ = consumes<std::vector<l1t::EtSum>>(TrackMETEmuInputTag);
+  if (HW_analysis_) {
+    TrackMETHWToken_ = consumes<std::vector<l1t::EtSum>>(TrackMETHWInputTag);
+  }
+}
+
+/////////////
+// DESTRUCTOR
+L1TkMETAnalyser::~L1TkMETAnalyser() {}
+
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc:116]:
```suggestion
```

File: L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissEmulatorProducer.cc [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+void L1TrackerEtMissEmulatorProducer::beginJob() {}
+
+void L1TrackerEtMissEmulatorProducer::endJob() {}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void L1TrackerEtMissEmulatorProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackerEtMissEmulatorProducer.cc:436]: Why not to include an actual parameter validation?

File: L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+#include "L1Trigger/L1TTrackMatch/interface/L1TkHTMissEmulatorProducer.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+using namespace l1t;
+
+class L1TkHTMissEmulatorProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1TkHTMissEmulatorProducer(const edm::ParameterSet&);
+  ~L1TkHTMissEmulatorProducer() override;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc:35]:
```suggestion
  ~L1TkHTMissEmulatorProducer() override = default;
```

File: L1Trigger/L1TTrackMatch/plugins/L1TkGlbMuonProducer.cc [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+    }
+  }  //over l1mus
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void L1TkGlbMuonProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TkGlbMuonProducer.cc:255]: Why not to allow an actual parameter validation?

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  unsigned int Sector;  //Phi sector
+  bool EtaSector;       //Positve or negative eta
+
+  float phi;  // Used to debug cos phi LUT
+};
+
+class L1TkEtMissEmuTrackTransform {
+public:
+  L1TkEtMissEmuTrackTransform() = default;
+  ~L1TkEtMissEmuTrackTransform(){};
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h:44]:
```suggestion
  ~L1TkEtMissEmuTrackTransform() = default;
```

**Review by perrotta — COMMENTED** [review]
A few sparse comments after a quick glance at the code

File: DataFormats/L1Trigger/src/classes_def.xml [hunk 15]
```diff
@@ -282,10 +282,22 @@
 
   <class name="edm::Ptr<l1t::L1Candidate>"/>
 
-  <class name="l1t::Vertex" ClassVersion="4">
-    <version ClassVersion="4" checksum="3366446278"/>
-    <version ClassVersion="3" checksum="514139083"/>
-  </class>
+  <class name="l1t::Vertex"/>
```
Comment (makortel) [inline DataFormats/L1Trigger/src/classes_def.xml:285]: Why remove the class versions but keep the class here?

File: DataFormats/L1Trigger/src/classes_def.xml [hunk 16]
```diff
@@ -282,10 +282,22 @@
...
   <class name="std::vector<l1t::Vertex>"/>
+  <class name="edm::Wrapper<std::vector<l1t::Vertex> >"/>
+
+  <class name="l1t::VertexWord" ClassVersion="10">
+    <version ClassVersion="10" checksum="3779192820"/>
+  </class>
+  <class name="std::vector<l1t::VertexWord>"/>
+  <class name="edm::Wrapper<std::vector<l1t::VertexWord> >"/>
+  <class name="l1t::VertexWord::VertexRef"/>
+  <class name="edm::Wrapper<l1t::VertexWord::VertexRef>"/>
```
Comment (makortel) [inline DataFormats/L1Trigger/src/classes_def.xml:295]: Do you intend to store a data product of just the `Ref`?

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h [hunk 17]
```diff
@@ -0,0 +1,77 @@
+#pragma once
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h:1]: Please use include guard instead.

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    return (ghweta % 2) ? glbeta_intern(ghweta + 1) : ghweta;
+  }
+
+  inline glbphi_intern makeGlbPhi(float phi) { return round(phi / PHI_LSB); }
+
+};  // namespace convert
+
+//Each individual box in the eta and phi dimension.
+//  Also used to store final cluster data for each zbin.
+struct EtaPhiBin {
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h:61]: There is already a class with this name in
https://github.com/cms-sw/cmssw/blob/522098e61e63196f7ac283966668910d88ecf188/L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h#L10

Defining a different class here violates the One Definition Rule and results in undefined behavior. In general such generic type names should be avoided in the global namespace.

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+  pt_intern pTtot;
+  l1t::TkJetWord::nt_t ntracks;
+  l1t::TkJetWord::nx_t nxtracks;
+  bool used;
+  glbphi_intern phi;  //average phi value (halfway b/t min and max)
+  glbeta_intern eta;  //average eta value
+};
+
+//store important information for plots
+struct MaxZBin {
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h:71]: Same here with
https://github.com/cms-sw/cmssw/blob/522098e61e63196f7ac283966668910d88ecf188/L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h#L22

File: L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+  std::vector<float>* m_SimMETphi;
+  std::vector<float>* m_EmuMETphi;
+  std::vector<float>* m_HwMETphi;
+
+  std::vector<float>* m_SimNtrk;
+  std::vector<float>* m_EmuNtrk;
+  std::vector<float>* m_HwNtrk;
+
+  edm::Service<TFileService> fs_;
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc:77]: Service handle should not be used as EDModule class member variables. It should be instantiated in the local scope of the member functions where it is used instead.

File: L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+#include "TGraphErrors.h"
+#include "TH1F.h"
+#include "TH2F.h"
+#include "TPad.h"
+#include "TProfile.h"
+#include "TTree.h"
+
+using namespace std;
+
+class L1TkMETAnalyser : public edm::one::EDAnalyzer<> {
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc:37]: The use of `TFileService` needs to be told to the framework with
```suggestion
class L1TkMETAnalyser : public edm::one::EDAnalyzer<edm::one::SharedResources> {
```
and calling
```cpp
usesResource(TFileService::kSharedResource);
```
in the constructor.

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc [hunk 22]
```diff
@@ -95,18 +99,30 @@ using namespace edm;
 //                          //
 //////////////////////////////
 
-class L1TrackObjectNtupleMaker : public edm::EDAnalyzer {
+class L1TrackObjectNtupleMaker : public edm::stream::EDAnalyzer<> {
```
Comment (makortel) [inline L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc:102]: This module uses `TFileService` which is not thread safe, and therefore the base class must be
```suggestion
class L1TrackObjectNtupleMaker : public edm::one::EDAnalyzer<edm::one::SharedResources> {
```
(plus the call to `usesResource(TFileService::kSharedResource);` in the constructor)

**Comment by cecilecaillol** [comment]
please test

File: DataFormats/L1TMuonPhase2/src/classes_def.xml [hunk 23]
```diff
@@ -0,0 +1,28 @@
+<lcgdict>
+ <selection>
+
+ <class name="l1t::MuonStub" ClassVersion="4">
+   <version ClassVersion="4" checksum="1164456369"/>
+   <version ClassVersion="3" checksum="939575486"/>
```
Comment (makortel) [inline DataFormats/L1TMuonPhase2/src/classes_def.xml:6]: Why start with version 4? The checksum fore version 3 is the same as for `l1t::TrackerMuon` and `l1t::SAMuon`, which looks suspicious.

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
@makortel @perrotta Your comments have been addressed.
@srimanob Can you please sign off again?

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
+Upgrade

The difference in PR test comes from Phase-2 DD4hep workflow which can be ignored.

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by qliphy** [comment]
This seems to cause [FWLite](https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/fwlite/slc7_amd64_gcc10/www/fri/12.4-fri-23/CMSSW_12_4_X_2022-03-25-2300) error:

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/fwlite/slc7_amd64_gcc10/CMSSW_12_4_X_2022-03-25-2300/DataFormats/L1Trigger

@cms-sw/core-l2 @cecilecaillol

**Comment by perrotta** [comment]
> This seems to cause [FWLite](https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/fwlite/slc7_amd64_gcc10/www/fri/12.4-fri-23/CMSSW_12_4_X_2022-03-25-2300) error:
> 
> https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/fwlite/slc7_amd64_gcc10/CMSSW_12_4_X_2022-03-25-2300/DataFormats/L1Trigger
> 
> @cms-sw/core-l2 @cecilecaillol

I think that the fix could be just removing 
```C++
#include "DataFormats/L1TMuonPhase2/interface/MuonStub.h"
#include "DataFormats/L1TMuonPhase2/interface/TrackerMuon.h"
#include "DataFormats/L1TMuonPhase2/interface/SAMuon.h"
```
from DataFormats/L1Trigger/src/classes.h
They would require an addition in the BuildfFile of DataFormats/L1TMuonPhase2, but it would have caused a recursive referencing between the two packages. In any case, inside DataFormats/L1Trigger/src/classes.h they are not needed any more, and therefore I imagine that the issue can be fixed by simply removing them from there.

PR #37368 implements that fix

**Comment by qliphy** [comment]
@smuzaffar may comment more, but it seems https://github.com/cms-sw/cmsdist/pull/7725 is fixing the [FWLite](https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/fwlite/slc7_amd64_gcc10/www/fri/12.4-fri-23/CMSSW_12_4_X_2022-03-25-2300) error

**Comment by perrotta** [comment]
Thank you @qliphy and @smuzaffar 
I'm closing https://github.com/cms-sw/cmssw/pull/37368 (which wasn't even correct...)

**Pull request merged**